liblumen_codegen/src/llvm.rs

```diff
@@ -0,0 +1,423 @@
+#![allow(dead_code)]
+
+mod enums;
+mod memory_buffer;
+mod target;
+
+use std::ffi::CString;
+use std::sync::{Once, ONCE_INIT};
+
+use llvm_sys::core::*;
+use llvm_sys::prelude::*;
+use llvm_sys::target::*;
+use llvm_sys::*;
+
+use codemap_diagnostic::{ColorConfig, Diagnostic, Emitter, Level};
+
+use super::CodeGenError;
+
+pub use self::enums::*;
+use self::target::{Target, TargetMachine};
+
+// Used to ensure LLVM is only initialized once
+static ONCE: Once = ONCE_INIT;
+
+/// Initialize LLVM internals
+pub fn initialize() {
+    ONCE.call_once(|| unsafe {
+        error_handling::LLVMEnablePrettyStackTrace();
+
+        LLVM_InitializeAllTargetInfos();
+        LLVM_InitializeAllTargets();
+        LLVM_InitializeAllTargetMCs();
+        LLVM_InitializeAllAsmParsers();
+        LLVM_InitializeAllAsmPrinters();
+
+        let registry = LLVMGetGlobalPassRegistry();
+        initialization::LLVMInitializeCore(registry);
+        initialization::LLVMInitializeTarget(registry);
+        initialization::LLVMInitializeCodeGen(registry);
+        initialization::LLVMInitializeAnalysis(registry);
+        initialization::LLVMInitializeTransformUtils(registry);
+    });
+}
+
+/// The handler function for LLVM diagnostics
+extern "C" fn diganostic_handler(info: LLVMDiagnosticInfoRef, ctx: *mut libc::c_void) {
+    let ctx: &mut Context = unsafe { &mut *(ctx as *mut Context) };
+    let severity = unsafe { LLVMGetDiagInfoSeverity(info) };
+    let description = unsafe { LLVMGetDiagInfoDescription(info) };
+    let d = Diagnostic {
+        level: severity_to_diagnostic_level(severity),
+        message: c_str_to_str!(description).to_string(),
+        code: None,
+        spans: Vec::new(),
+    };
+    ctx.emitter.emit(&vec![d]);
+}
+
+fn severity_to_diagnostic_level(severity: LLVMDiagnosticSeverity) -> Level {
+    match severity {
+        LLVMDiagnosticSeverity::LLVMDSError => Level::Error,
+        LLVMDiagnosticSeverity::LLVMDSWarning => Level::Warning,
+        LLVMDiagnosticSeverity::LLVMDSRemark => Level::Help,
+        LLVMDiagnosticSeverity::LLVMDSNote => Level::Note,
+    }
+}
+
+pub struct Module {
```

Comment:
Does this `Module` map to an Erlang module or is it more like a compilation unit in C++/Rust?

Reply:
This was included with the changes from Paul I think?